src/ray/core_worker/core_worker.cc

```diff
@@ -3722,47 +3755,55 @@ Status CoreWorker::GetAndPinArgsForExecutor(const TaskSpecification &task,
-      borrowed_ids->push_back(arg_id);
-      // We need to put an OBJECT_IN_PLASMA error here so the subsequent call to Get()
-      // properly redirects to the plasma store.
-      // NOTE: This needs to be done after adding reference to reference counter
-      // otherwise, the put is a no-op.
-      if (!options_.is_local_mode) {
-        RAY_UNUSED(memory_store_->Put(RayObject(rpc::ErrorType::OBJECT_IN_PLASMA),
-                                      task.ArgId(i)));
-      }
+    if (task.ArgByRef(i) && task.ArgIsErrorType(i, rpc::ErrorType::OBJECT_IN_PLASMA)) {
+        const auto &arg_ref = task.ArgRef(i);
+        const auto arg_id = ObjectID::FromBinary(arg_ref.object_id());
+        by_ref_plasma_ids.insert(arg_id);
+        auto it = by_ref_indices.find(arg_id);
+        if (it == by_ref_indices.end()) {
+          by_ref_indices.emplace(arg_id, std::vector<size_t>({i}));
+        } else {
+          it->second.push_back(i);
+        }
+        arg_refs->at(i) = arg_ref;
+        // Pin all args passed by reference for the duration of the task.  This
+        // ensures that when the task completes, we can retrieve metadata about
+        // any borrowed ObjectIDs that were serialized in the argument's value.
+        RAY_LOG(DEBUG).WithField(arg_id) << "Incrementing ref for argument ID";
+        reference_counter_->AddLocalReference(arg_id, task.CallSiteString());
+        // Attach the argument's owner's address. This is needed to retrieve the
+        // value from plasma.
+        reference_counter_->AddBorrowedObject(
+            arg_id, ObjectID::Nil(), task.ArgRef(i).owner_address());
+        borrowed_ids->push_back(arg_id);
+        // We need to put an OBJECT_IN_PLASMA error here so the subsequent call to Get()
+        // properly redirects to the plasma store.
+        // NOTE: This needs to be done after adding reference to reference counter
+        // otherwise, the put is a no-op.
+        if (!options_.is_local_mode) {
+          RAY_UNUSED(memory_store_->Put(RayObject(rpc::ErrorType::OBJECT_IN_PLASMA),
+                                        task.ArgId(i)));
+        }
     } else {
+      std::shared_ptr<Buffer> data = nullptr;
+      std::shared_ptr<Buffer> metadata = nullptr;
       // A pass-by-value argument.
-      std::shared_ptr<LocalMemoryBuffer> data = nullptr;
       if (task.ArgDataSize(i) != 0u) {
         data = std::make_shared<LocalMemoryBuffer>(const_cast<uint8_t *>(task.ArgData(i)),
                                                    task.ArgDataSize(i));
+      } else if (task.ArgByRef(i)) {
+        RAY_CHECK(task.ArgIsErrorType(i, rpc::ErrorType::OBJECT_IN_ACTOR));
+        const auto it = in_actor_dependencies_metadata.find(task.ArgId(i));
+        RAY_CHECK(it != in_actor_dependencies_metadata.end());
+        data = it->second;
+        metadata = std::make_shared<LocalMemoryBuffer>(const_cast<uint8_t *>(task.ArgMetadata(i)),
```

Comment:
`RayActorObjectMetadata` this is passed to `task_execution_handler` by `args`, and then Cython can use the src_rank from metadata to call NCCL recv.